controllers/workspace/finalize.go

```diff
@@ -0,0 +1,213 @@
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/controllers/workspace/provision"
+	"github.com/devfile/devworkspace-operator/internal/images"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/go-logr/logr"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	pvcClaimMountPath   = "/tmp/devworkspaces/"
+	cleanupCommandFmt   = "rm -rf %s"
+	pvcCleanupFinalizer = "storage.controller.devfile.io"
+)
+
+var (
+	cleanupJobCompletions  = int32(1)
+	cleanupJobBackoffLimit = int32(3)
+)
+
+func (r *DevWorkspaceReconciler) setFinalizer(ctx context.Context, workspace *v1alpha2.DevWorkspace) (ok bool, err error) {
+	if !isFinalizerNecessary(workspace) || hasFinalizer(workspace) {
+		return true, nil
+	}
+	workspace.SetFinalizers(append(workspace.Finalizers, pvcCleanupFinalizer))
+	return false, r.Update(ctx, workspace)
+}
+
+func (r *DevWorkspaceReconciler) finalize(ctx context.Context, log logr.Logger, workspace *v1alpha2.DevWorkspace) error {
+	if !hasFinalizer(workspace) {
+		return nil
+	}
+	// Need to make sure Deployment is cleaned up before starting job to avoid mounting issues for RWO PVCs
+	wait, err := provision.DeleteWorkspaceDeployment(ctx, workspace, r.Client)
+	if err != nil {
+		return err
+	}
+	if wait {
+		return nil
+	}
+	specJob, err := r.getSpecCleanupJob(workspace)
+	if err != nil {
+		return err
+	}
+	clusterJob, err := r.getClusterCleanupJob(ctx, workspace)
+	if err != nil {
+		return err
+	}
+	if clusterJob == nil {
+		err := r.Create(ctx, specJob)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		return nil
+	}
+	if !equality.Semantic.DeepDerivative(specJob.Spec, clusterJob.Spec) {
+		err := r.Patch(ctx, specJob, client.Merge)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	for _, condition := range clusterJob.Status.Conditions {
+		if condition.Status != corev1.ConditionTrue {
+			continue
+		}
+		switch condition.Type {
+		case batchv1.JobComplete:
+			log.Info("PVC clean up job successful; clearing finalizer")
+			clearFinalizer(workspace)
+			return r.Update(ctx, workspace)
+		case batchv1.JobFailed:
+			log.Error(fmt.Errorf("PVC clean up job failed: message %q", condition.Message),
```

Comment:
Does the message is really detailed? Or it just says - `container exited with 1 code`
If so, please consider propagating job container logs into devworkspace logs

Reply:
The pod gets a message like `exited with 1 code` but the job gets something referring to "backoff limit reached". I don't know a good way to propagate useful information -- the logs will generally just be something like "permission denied" from the pods, and I don't know if there's an easy way to get relevant statuses (if a pod fails for any other reason).

I'll poke at it and see if I can figure something out.

Reply:
My current workaround is to not clear finalizers when they fail so that you can look at the pods/their logs directly to figure out what went wrong.